src/stylefunction.js

```diff
@@ -975,6 +1027,16 @@ export function stylefunction(
             circleStrokeWidth;
           iconImg = iconImageCache[cache_key];
           if (!iconImg) {
+            // the mapbox style spec does not support decluttering circles.
+            // but as openlayers does, we can support it as "extension"
+            const declutterMode = getDeclutterMode(
+              layer,
+              'circle-allow-overlap',
+              'circle-ignore-placement',
+              zoom,
+              f,
+              functionCache
+            );
```

Comment:
Please do not add style properties that the spec does not support. Instead, there is something else to fix: Currently, we treat circles different than mapbox-gl-js, because we declutter them, which mapbox-gl-js does not. So a better change would be to change circle rendering to never declutter (`declutterMode: 'none'`).

When you change this, the signature of `getDeclutterMode()` will also be simpler, because it does not need the 2nd and 3rd argument any more.

Reply:
The nice thing about the mapbox styles is the ability to configure the styles from a configuration file (which is the reason we want to use it for configuring openlayers layers, too).
However, there are openlayers features that are not supported by mapbox, like decluttering circles, for which it would be nice to have a configuration possibility, here with additional properties that don't hurt.

I would like to keep the properties, but can remove them, if you insist.
To confirm to the mapbox behaviour, the default for the circles will be declutter: none.

However, as before the PR, the style function only works correctly when applied to a layer that has declutter set correctly depending on the style's text-allow-overlap. When using apply for the whole map/mapbox style sheet, all layers are created with declutter = true. This is not solved with this PR.

Reply:
The problem is: if you leave these undocumented properties in, they are not documented anywhere. I use ol-mapbox-style too for configuring styles from static objects, and for customisations I use layer metadata and a style function wrapper that modifies the returned styles. If `declutterMode` had a setter on `ol/style/Image`, this would be very simple in your case too. We could make it even simpler by providing a style customization hook, something like
```js
/**
 * @param {import("ol/style/Style.js").default} style Style.
 * @param {Object} glLayer Item from the Mapbox style's `layers` array.
 * @param {Object} glSource Item from the Mapbox style's `sources` object.
customizeStyle(style, glLayer, glSource) {
  if (glLayer.metadata.declutter-circle && style.getImage() instanceof CircleStyle) {
    style.getImage().setDeclutterMode(glLayer.metadata.declutter-circle)
  }
}
```
That function could become a new config option, to be configured in the `options` object.